pkg/ccl/streamingccl/streamproducer/event_stream.go

```diff
@@ -255,97 +179,93 @@ func (s *eventStream) Close(ctx context.Context) {
-	}
+func (s *eventStream) onInitialScanDone(ctx context.Context) {
+	// We no longer expect concurrent onValue calls so we can remove the mu.
+	s.addMu = nil
 }
 
-func (s *eventStream) onInitialScanSpanCompleted(ctx context.Context, sp roachpb.Span) error {
-	checkpoint := kvpb.RangeFeedCheckpoint{
-		Span:       sp,
-		ResolvedTS: s.spec.InitialScanTimestamp,
-	}
-	select {
-	case <-ctx.Done():
-		return ctx.Err()
-	case s.eventsCh <- kvcoord.RangeFeedMessage{
-		RangeFeedEvent: &kvpb.RangeFeedEvent{Checkpoint: &checkpoint},
-	}:
-		log.VInfof(ctx, 1, "onSpanCompleted: %s@%s", checkpoint.Span, checkpoint.ResolvedTS)
-		return nil
-	}
+func (s *eventStream) onValue(ctx context.Context, value *kvpb.RangeFeedValue) {
+	// During initial-scan we expect concurrent onValue calls from the parallel
+	// scan workers, but once the initial scan ends the mu will be nilled out and
+	// we can avoid the locking overhead here.
+	if s.addMu != nil {
+		s.addMu.Lock()
+		defer s.addMu.Unlock()
+	}
+	s.seb.addKV(roachpb.KeyValue{Key: value.Key, Value: value.Value})
+	s.setErr(s.maybeFlushBatch(ctx))
 }
 
 func (s *eventStream) onSSTable(
 	ctx context.Context, sst *kvpb.RangeFeedSSTable, registeredSpan roachpb.Span,
 ) {
-	select {
-	case <-ctx.Done():
-	case s.eventsCh <- kvcoord.RangeFeedMessage{
-		RangeFeedEvent: &kvpb.RangeFeedEvent{SST: sst},
-		RegisteredSpan: registeredSpan,
-	}:
-		log.VInfof(ctx, 1, "onSSTable: %s@%s with registered span %s",
-			sst.Span, sst.WriteTS, registeredSpan)
+	if s.setErr(s.addSST(sst, registeredSpan)) {
+		return
 	}
+	s.setErr(s.maybeFlushBatch(ctx))
 }
 
 func (s *eventStream) onDeleteRange(ctx context.Context, delRange *kvpb.RangeFeedDeleteRange) {
-	select {
-	case <-ctx.Done():
-	case s.eventsCh <- kvcoord.RangeFeedMessage{RangeFeedEvent: &kvpb.RangeFeedEvent{DeleteRange: delRange}}:
-		log.VInfof(ctx, 1, "onDeleteRange: %s@%s", delRange.Span, delRange.Timestamp)
+	s.seb.addDelRange(*delRange)
+	s.setErr(s.maybeFlushBatch(ctx))
+}
+
+func (s *eventStream) maybeCheckpoint(ctx context.Context, advanced bool, frontier rangefeed.VisitableFrontier) {
+	if s.addMu != nil {
+		s.addMu.Lock()
+		defer s.addMu.Unlock()
+	}
+	age := timeutil.Since(s.lastCheckpointTime)
+	if (advanced && age > s.spec.Config.MinCheckpointFrequency) || (age > 2*s.spec.Config.MinCheckpointFrequency) {
+		s.sendCheckpoint(ctx, frontier)
 	}
 }
 
-// makeCheckpoint generates checkpoint based on the frontier.
-func makeCheckpoint(f span.Frontier) (checkpoint streampb.StreamEvent_StreamCheckpoint) {
-	f.Entries(func(sp roachpb.Span, ts hlc.Timestamp) (done span.OpResult) {
-		checkpoint.ResolvedSpans = append(checkpoint.ResolvedSpans, jobspb.ResolvedSpan{
-			Span:      sp,
-			Timestamp: ts,
-		})
+func (s *eventStream) sendCheckpoint(ctx context.Context, frontier rangefeed.VisitableFrontier) {
+	if err := s.flushBatch(ctx); err != nil {
+		return
+	}
+
+	spans := make([]jobspb.ResolvedSpan, 0, s.lastCheckpointLen)
```

Comment:
huh. I wonder if c2c would be more performant if we figured out a way to not send the whole frontier over the wire every time we sent a checkpoint message. A problem for another day.